azkaban-exec-server/src/main/java/azkaban/execapp/ProjectCacheCleaner.java

```diff
@@ -0,0 +1,146 @@
+      public boolean accept(final File dir, final String name) {
+        return name.matches(this.pattern);
+      }
+    })) {
+      if (project.exists() && project.isDirectory()) {
+        projects.add(project.toPath());
+      } else {
+        log.debug("project {} doesn't exist or is non-dir.", project.getName());
+      }
+    }
+    return projects;
+  }
+
+  private List<ProjectDirectoryMetadata> loadAllProjects() {
+    final List<ProjectDirectoryMetadata> allProjects = new ArrayList<>();
+    for (final Path project : this.loadAllProjectDirs()) {
+      try {
+        final String fileName = project.getFileName().toString();
+        final int projectId = Integer.parseInt(fileName.split("\\.")[0]);
+        final int versionNum = Integer.parseInt(fileName.split("\\.")[1]);
+
+        final ProjectDirectoryMetadata projectDirMetadata =
+            new ProjectDirectoryMetadata(projectId, versionNum, project.toFile());
+
+        projectDirMetadata.setDirSizeInByte(
+            FlowPreparer.calculateDirSizeAndSave(projectDirMetadata.getInstalledDir()));
+        projectDirMetadata.setLastAccessTime(
+            Files.getLastModifiedTime(Paths.get(projectDirMetadata.getInstalledDir().toString(),
+                FlowPreparer.PROJECT_DIR_SIZE_FILE_NAME)));
+        allProjects.add(projectDirMetadata);
+      } catch (final Exception e) {
+        log.warn("error while loading project dir metadata for project {}",
+            project.getFileName(), e);
+      }
+    }
+    return allProjects;
+  }
+
+  /**
+   * @return sum of the size of all project dirs
+   */
+  private long getProjectDirsTotalSizeInBytes(final List<ProjectDirectoryMetadata> allProjects) {
+    final long totalSizeInBytes = allProjects.stream()
+        .mapToLong(ProjectDirectoryMetadata::getDirSizeInByte).sum();
+    return totalSizeInBytes;
+  }
+
+  private void deleteLeastRecentlyUsedProjects(long sizeToFreeInBytes,
+      final List<ProjectDirectoryMetadata> projectDirMetadataList) {
+    // Sort projects by last reference time in ascending order
+    projectDirMetadataList.sort(Comparator.comparing(ProjectDirectoryMetadata::getLastAccessTime));
+    for (final ProjectDirectoryMetadata proj : projectDirMetadataList) {
+      if (sizeToFreeInBytes > 0) {
+        // Delete the project directory even if flow within is running. It's OK to
+        // delete the directory since execution dir is HARD linked to project dir.
+        log.debug("deleting project {}", proj);
+        FileIOUtils.deleteDirectory(proj.getInstalledDir());
+        sizeToFreeInBytes -= proj.getDirSizeInByte();
+      } else {
+        break;
+      }
+    }
+  }
+
+  /**
+   * Deleting least recently accessed project dirs when there's no room to accommodate new project.
+   */
+  void deleteProjectDirsIfNecessary(final long newProjectSizeInBytes) {
+    final long start = System.currentTimeMillis();
+    final List<ProjectDirectoryMetadata> allProjects = loadAllProjects();
```

Comment:
Do you know how long this takes? Thanks for adding the logging, to get the information. If this is expensive, the executor could also store the information, and update as directories are created and used.

Reply:
i don't have a number yet, that's why i added logging here to see if it worths further optimization.